### DISCUSSION TITLE
How to use a local external library in a Next.js app while developing?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#61800] [Help]
### Summary

What is the approach to using a _local_ external library (a TypeScript library which I'm developing concurrently) inside a Next.js app? How do you wire this up? I am using `pnpm`, can I `pnpm link` them, or do I need to symlink in a sort of way? What is the recommended approach?

I tried pnpm linking with and without this in my next.config.js, but both result in a host of errors of "module not found".

```
webpack: (config, options) => {
  config.resolve = {
    ...config.resolve,
    alias: {
      ...config.resolve.alias,
      '@myorg/myrepo': path.resolve(
        __dirname,
        './node_modules/@myorg/myrepo',
      ),
    },
  }
}
```

**Comment by taig** [comment]
I got it working by adding the dependency with as an absolute path and transpiling the package by Next.js.

**package.json**
```
{
  "dependencies": {
    "@my/app": "file:/absolute/path/to/my/app"
  }
}
```

**next.config.mjs**
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  transpilePackages: ["@my/app"],
}

export default nextConfig
```

**Reply by lancejpollard** [reply]
Nice! Will have to give this a try.

**Reply by rushiiMachine** [reply]
Something to note is that a `link:` dependency did not end up working for me, I had to switch to `file:`

**Comment by deflexable** [comment]
@taig hard coding with absolute path will be messy when pushing the code across multiple systems and having to manually change it per system

i tried using relative path like `file:/../path/to/app` but it stop working and brings `Module not found` error.

If anyone found a more suitable solution, kindly let us know

**Comment by CodeByKhaleel** [comment]
You can use a local package during development using either `npm link`, `pnpm link`, or workspace-based monorepo setup.

For quick local development:

```bash
cd your-local-library
npm link
```

Then inside your Next.js app:

```bash
npm link your-library-name
```

If you're using TypeScript, make sure:

* `main` and `types` fields are properly set in the library `package.json`
* the library is built before importing
* `transpilePackages` is configured in `next.config.js` if needed

Example:

```js
const nextConfig = {
  transpilePackages: ['your-library-name'],
}

module.exports = nextConfig
```

For long-term development, Turborepo or pnpm workspaces are usually cleaner than `npm link`.

**Comment by deflexable** [comment]
was able to fix the issue by prefixing with `./` 

something like `file:./../path/to/pkg` instead of `file:../path/to/pkg`

**Comment by hwajongpark** [comment]
For the portable version, I would use a workspace dependency if you can, and a relative `file:` dependency if you cannot.

**Best long-term setup: workspace package**

```jsonc
// package.json at the repo root
{
  "workspaces": ["apps/web", "packages/my-app"]
}
```

```jsonc
// apps/web/package.json
{
  "dependencies": {
    "@my/app": "workspace:*"
  }
}
```

```js
// apps/web/next.config.js
const nextConfig = {
  transpilePackages: ["@my/app"],
}

module.exports = nextConfig
```

That keeps the path portable across machines and lets Next compile the local package source. This is exactly the case `transpilePackages` is meant for: local packages / monorepo packages that Next should transpile and bundle.

**If the package is outside the workspace**

Use a relative `file:` spec, but watch the slash:

```jsonc
{
  "dependencies": {
    "@my/app": "file:../my-app"
    // or "file:./../my-app"
  }
}
```

Do **not** use `file:/../my-app`. The slash after `file:` makes it an absolute path from the filesystem root, which is why it breaks portability. `file:/absolute/path/to/my/app` works locally, but it is machine-specific.

After changing this, remove the old installed package/lock entry and reinstall so your package manager rewrites the symlink/resolution. Also make sure the local package has a real `name` field matching the import, plus `exports`/`main`/`types` as appropriate.

Docs: https://nextjs.org/docs/app/api-reference/config/next-config-js/transpilePackages
